cgi-bin/DW/Controller/EditIcons.pm

```diff
@@ -0,0 +1,391 @@
+# the same terms as Perl itself. For a copy of the license, please reference
+# 'perldoc perlartistic' or 'perldoc perlgpl'.
+#
+
+package DW::Controller::EditIcons;
+
+use strict;
+use v5.10;
+use Log::Log4perl;
+my $log = Log::Log4perl->get_logger( __PACKAGE__ );
+
+use File::Type;
+
+
+sub update_userpics {
+    my ( $ML, $POST, $errors, $userpicsref, $u, $err ) = @_;
+    my @userpics = @$userpicsref;
+    my $display_rename = LJ::is_enabled( "icon_renames" ) ? 1 : 0 ;
+
+    # form being posted isn't multipart, since we were able to read from %POST
+    unless (LJ::check_form_auth()) {
+        return $err->($ML->{'error.invalidform'});
+    }
+
+    my @delete; # userpic objects to delete
+    my @inactive_picids;
+    my %picid_of_kwid;
+    my %used_keywords;
+
+    # we need to count keywords based on what the user provided, in order
+    # to find duplicates. $up->keywords doesn't work, because re-using a
+    # keyword will remove it from the other userpic without our knowing
+    my $count_keywords = sub {
+        my $kwlist = shift;
+        $used_keywords{$_}++ foreach split(/,\s*/, $kwlist);
+    };
+
+    foreach my $up (@userpics) {
+
+        my $picid = $up->id;
+
+        # delete this pic
+        if ($POST->{"delete_$picid"}) {
+            push @delete, $up;
+            next;
+        }
+
+        # we're only going to modify keywords/comments on active pictures
+        if ($up->inactive || $POST->{"pic_inactive_$picid"}) {
+            # use 'orig' because we don't POST disabled fields
+            $count_keywords->($POST->{"kw_orig_$picid"});
+            next;
+        }
+
+        $count_keywords->($POST->{"kw_$picid"});
+
+        # only modify if changing the data, make sure not colliding with other edits, etc
+        if ($POST->{"kw_$picid"} ne $POST->{"kw_orig_$picid"}) {
+            my $kws = $POST->{"kw_$picid"};
+
+            if ( $POST->{"rename_keyword_$picid"} ) {
+                if ( $display_rename && $u->userpic_have_mapid ) {
+                    eval {
+                       $up->set_and_rename_keywords($kws, $POST->{"kw_orig_$picid"});
+                    } or push @$errors, $@->as_html;
+                } else {
+                    push @$errors, $ML->{'.label.rename.disabled'};
+                }
+            } else {
+                eval {
+                    $up->set_keywords($kws);
+                } or push @$errors, $@->as_html;
+            }
+        }
+
+        eval {
+            $up->set_comment ($POST->{"com_$picid"})
+                unless $POST->{"com_$picid"} eq $POST->{"com_orig_$picid"};
+        } or push @$errors, $@;
+
+        eval {
+            $up->set_description ($POST->{"desc_$picid"})
+                unless $POST->{"desc_$picid"} eq $POST->{"desc_orig_$picid"};
+        } or push @$errors, $@;
+
+    }
+
+    foreach my $kw (keys %used_keywords) {
+        next unless $used_keywords{$kw} > 1;
+        push @$errors, BML::ml('.error.keywords', {ekw => $kw});
```

Comment:
Oh, OK! That means it works for now because the calling context is still BML, but it will suddenly stop working when the rest of the page is converted, so that's still a long-term sadness.

Here's an example of a the sort of thing I was talking about - this example uses a variable instead of a function: https://github.com/dreamwidth/dw-free/blob/develop/cgi-bin/DW/Controller/Settings.pm#L51